esr/thing_test.go

```diff
@@ -0,0 +1,729 @@
+	}
+
+	for _, s := range expectedServices {
+		if _, ok := services[s]; !ok {
+			t.Errorf("Expected service '%s' to be present", s)
+		}
+	}
+}
+
+// ------------------------------------------------ //
+// Help functions and structs to test newResource()
+// ------------------------------------------------ //
+
+// Create a error reader to break json.Unmarshal()
+type errReader int
+
+var errBodyRead error = fmt.Errorf("bad body read")
+
+func (errReader) Read(p []byte) (n int, err error) {
+	return 0, errBodyRead
+}
+func (errReader) Close() error {
+	return nil
+}
+
+func createConfAssetBrokenTraits() usecases.ConfigurableAsset {
+	brokenTrait, _ := json.Marshal(errReader(0))
+	uac := usecases.ConfigurableAsset{
+		Name:     "testRegistrar",
+		Details:  map[string][]string{"testDetail": {"detail1", "detail2"}},
+		Services: []components.Service{},
+		Traits:   []json.RawMessage{json.RawMessage(brokenTrait)},
+	}
+	return uac
+}
+
+func createConfAssetMultipleTraits() usecases.ConfigurableAsset {
+	uac := usecases.ConfigurableAsset{
+		Name:     "testRegistrar",
+		Details:  map[string][]string{"testDetail": {"detail1", "detail2"}},
+		Services: []components.Service{},
+		Traits:   []json.RawMessage{json.RawMessage(`{"recCount": 0}`), json.RawMessage(`{"leading": false}`)},
+	}
+	return uac
+}
+
+func createTestSystem() components.System {
+	ctx := context.Background()
+	sys := components.NewSystem("testsys", ctx)
+	sys.Husk = &components.Husk{
+		Description: " is for testing purposes",
+		Certificate: "ABCD",
+		Details:     map[string][]string{"Developer": {"Arrowhead"}},
+		ProtoPort:   map[string]int{"https": 0, "http": 8870, "coap": 0},
+		InfoLink:    "https://for.testing.purposes",
+	}
+	leadingRegistrar := &components.CoreSystem{
+		Name: components.ServiceRegistrarName,
+		Url:  "https://leadingregistrar:1234",
+	}
+	orchestrator := &components.CoreSystem{
+		Name: "orchestrator",
+		Url:  "https://orchestator:1234",
+	}
+	sys.CoreS = []*components.CoreSystem{
+		leadingRegistrar,
+		orchestrator,
+	}
+	return sys
+}
+
+type newResourceParams struct {
+	setup     func() components.System
+	confAsset func() usecases.ConfigurableAsset
+	testCase  string
+}
+
+func TestNewResource(t *testing.T) {
+	params := []newResourceParams{
+		{
+			func() (sys components.System) { return createTestSystem() },
+			func() (confAsset usecases.ConfigurableAsset) { return createConfAssetBrokenTraits() },
+			"Case: unmarshal traits fails",
+		},
+		{
+			func() (sys components.System) { return createTestSystem() },
+			func() (confAsset usecases.ConfigurableAsset) { return createConfAssetMultipleTraits() },
+			"Case: confAsset has multiple traits",
+		},
+	}
+
+	for _, c := range params {
+		sys := c.setup()
+		uac := c.confAsset()
+
+		ua, shutdown := newResource(uac, &sys)
+		shutdown()
+		if ua.GetName() != "testRegistrar" {
+			t.Errorf("Name mismatch, expected '%s' got '%s'", uac.Name, ua.GetName())
+		}
+	}
+}
```

Comment:
The `newResource` is just wonky atm, skip testing it even if it's doing more than just setting fields. Not worth chasing coverage percent for a function that risk breaking in a month.